crates/sui-framework/sources/governance/voting_power.move

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+module sui::voting_power {
+    use sui::validator::Validator;
+    use std::vector;
+    use sui::validator;
+    use sui::math;
+
+    struct VotingPowerInfo has drop {
+        validator_index: u64,
+        voting_power: u64,
+    }
+
+    /// Set total_voting_power as 10_000 by convention. Individual voting powers can be interpreted
+    /// as easily understandable basis points (e.g., voting_power: 100 = 1%, voting_power: 1 = 0.01%) rather than
+    /// opaque quantities whose meaning changes from epoch to epoch as the total amount staked shifts.
+    /// Fixing the total voting power allows clients to hardcode the quorum threshold and total_voting power rather
+    /// than recomputing these.
+    const TOTAL_VOTING_POWER: u64 = 10_000;
+
+    /// Quorum threshold for our fixed voting power--any message signed by this much voting power can be trusted
+    /// up to BFT assumotions
+    const QUORUM_THRESHOLD: u64 = 6_667;
+
+    // Cap voting power of an individual validator at 10%.
+    // TODO: determine what this should be
+    const MAX_VOTING_POWER: u64 = 1_000;
+
+    const ETotalPowerMismatch: u64 = 1;
+    const ERelativePowerMismatch: u64 = 2;
+    const EVotingPowerOverThreshold: u64 = 3;
+
+    /// Set the voting power of all validators. The total stake of all validators is provided in `total_stake`.
+    /// threshold_pct is a percentage threshold of max voting power that we want to cap on. If threshold_pct is 10,
+    /// then we want to cap the voting power at 10%.
+    public fun set_voting_power(validators: &mut vector<Validator>, total_stake: u64) {
```

Comment:
We need to actually call this from `validator_set` at the end of epoch change.

Reply:
That will happen in a separate PR (Sam's original PR)